README.md
#jQThumb v2.3.6

Create thumbnails from images proportionally. On top of that, this is alaso a lazy-load plugin, which even works on IE6 from jQuery >=v1.3 or Zepto (with zepto-data plugin) >=v1.1.3.

![screenshot](screenshots/screenshot.jpg?raw=true "jQThumb Screenshot")

#USAGE
```html
<!DOCTYPE html>
<html lang="en">
    <body>
        <div style="width: 100%; height: 400px;">
            <img src="path/picture.jpg" class="example1" />
        </div>
        <div style="width: 400px; height: 400px;">
            <img src="path/picture.jpg" class="example2" />
        </div>
        <button id="kill">Kill</button>
        <button id="kill-all">Kill All</button>
    </body>
    <script type="text/javascript" src="scripts/jquery.min.js"></script>
    <script type="text/javascript" src="scripts/jqthumb.min.js"></script>
    <script type="text/javascript">
        $(function(){

            // plugin initialization
            $('img').jqthumb({
                classname      : 'jqthumb',             // class name. DEFUALT IS jqthumb
                width          : '100%',                // new image width after cropping. DEFAULT IS 100px.
                height         : '100%',                // new image height after cropping. DEFAULT IS 100px.
                position       : {
                    x : '50%',                          // x position of the image. DEFAULT is 50%. 50% also means centerize the image.
                    y : '50%'                           // y position of the image. DEFAULT is 50%. 50% also means centerize the image.
                },
                source         : 'src',                 // to specify the image source attribute. DEFAULT IS src.
                show           : false,                 // TRUE = show immediately after processing. FALSE = do not show it. DEFAULT IS TRUE.
                renderPosition : 'before',              // available: "before" and "after".
                onDemand       : false,                 // TRUE = load image only when scolling position has reached the DOM
                onDemandEvent  : 'scroll',              // available: "scroll", "click", "mouseenter". DEFAULT IS "scroll"
                threshold      : 0,                     // used when "onDemand" is set to true AND "onDemandEvent" is set to "scroll". Eg. Start loading the image 200px before scolling position reaches the DOM. DEFUALT IS 0
                responsive     : 20,                    // used by older browsers only. 0 to disable. DEFAULT IS 20
                zoom           : 1,                     // zoom the output, 2 would double of the actual image size. DEFAULT IS 1
                method         : 'auto',                // 3 methods available: "auto", "modern" and "native". DEFAULT IS auto
                reinit         : true,                  // TRUE = to re-init when images is re-initialized for the second time. FALSE = nothing would happen.
                error          : function(dom, imgUrl){ // callback on error, returns image url
                    console.log(dom, ' with its url "' + imgUrl + '" is invalid.');
                }
                before         : function(oriImage){    // callback before each image starts processing.
                    alert("I'm about to start processing now...");
                },
                after          : function(imgObj){      // callback when each image is cropped.
                    console.log(imgObj);
                },
                done           : function(imgArray){    // callback when all images are cropped.
                    for(i in imgArray){
                        $(imgArray[i]).fadeIn();
                    }
                }
            });

            // kill command
            $('#kill').click(function(){
                $('.your-dom').jqthumb('kill');
            });

            // kill all command
            $('#destroy-all').click(function(){
                $.jqthumb('killall');
            });
        });
    </script>
</html>
```

##INTRODUCTION
This is a plugin helps creating thumbnails proportionally from images. As many of you may know that `background-size: cover;` would solve most of the major issues when dealing with thumbnails. But `background-size: cover;` does not work in older browsers like IE6, 7 and 8 therefore this is one of the reasons why this plugin was built.

Ever wonder how to support full-width billboard in older browsers that works the same as modern browser? This plugin helps exactly in this. Never assume this plugin only generates thumbnails, in fact this works perfectly with big images like billboards. On top of that, the plugin also comes with lazy-load feature.

##BOWER
`bower install jqthumb`

##DEMO
http://pakcheong.github.io/jqthumb/

##DEFAULT OPTIONS
```javascript
$.fn.jqthumb.defaults = {
    classname      : 'jqthumb',
    width          : 100,
    height         : 100,
    position       : { x: '50%', y: '50%' },
    source         : 'src',
    responsive     : 20,
    zoom           : 1,
    show           : true,
    renderPosition : 'before',
    onDemand       : false,
    onDemandEvent  : 'scroll',
    threshold      : 0,
    method         : 'auto',
    reinit         : true,
    error          : function(){},
    before         : function(){},
    after          : function(){},
    done           : function(){}
};
```

##OPTION REFERENCES

####source
The image path attribute of the HTML tag. The source for `<img src="path/image.jpg" />` would be `src`.
```javascript
$('img').jqthumb({
    source : 'attr-src' // DEFAULT: src
});
```

####classname
The class name of the generated thumbnail. This is useful when you want to attach extra stylings to the thumbnail.
```javascript
$('img').jqthumb({
    classname : 'jqthumb-class-name' // DEFAULT: jqthumb
});
```

####width & height
The width of the generated thumbnail. This accepts both integer and string data types. Integer input would mean the width of the thumbnail is in pixel rather than percentage and vice versa. You may also set both to `auto` which means you're defining the ouput same as the actual resolution of the file. **Note: if you define width and/or height in percentage, make sure you have a container with width and/or height defined in pixels.**
```javascript
$('img').jqthumb({
    width  : 200,   // DEFAULT: 100
    height : '100%' // DEFAULT: 100
});
```

####position
This has to be defined as an object with **x** and **y** as its keys. **y** would be used to adjust the top-bottom position of the thumbnail and **x** adjusts left and right. **Note: both position.x and position.y must be within the range of the defined `width` and `height` respectively. If you are defining `position.x` and/or `position.y` in percenrage values instead, make sure it is within 0 to 100%**
```javascript
$('img').jqthumb({
    position: {
        x : 20,   // DEFAULT: '50%'
        y : '30%' // DEFAULT: '50%'
    }
});
```

####show
Whether to show the thumbnail right after processing.
```javascript
$('img').jqthumb({
    show : false // DEFAULT: true
});
```

####renderPosition
Render image whether before or after the selected DOM.
```javascript
$('img').jqthumb({
    renderPosition : 'after' // DEFAULT: 'before'
});
```

####onDemand / onDemandEvent / threshold
Asign an event to tell when to load the images. For eg., setting the event to "scroll" is a common action as you might want to load the images only when its DOM is within the viewport. Therefore, images will not start loading/processing until the scrolling position has reached the DOM. This is good when you have a lot of images on the page but user doesn't actually look through the entire site, so no point loading all at once.

`threshold`: Used only when `onDemand` is enabled AND `onDemandEvent` is set to "scroll". The scroll event will be triggered once the scrolling position has reached the DOM. For eg., you might want to load the images without users knowing it, so you will need to set `threshold` to maybe "200" which means images will start loading 200PX before (all directions including top, left, bottom, right) before scrolling position reaches the DOM.

`onDemandEvent`: has three possible inputs, "scroll", "click" and "mouseenter". Clicking and mouse hovering events mean that the images will only start loading when selected event is being triggered. For eg., setting `onDemandEvent` to "mouseenter" will lead image to start loading when users move the mouse cursor over it.
```javascript
$('img').jqthumb({
    onDemand      : true,     // DEFAULT: false
    onDemandEvent : 'scroll', // DEFAULT: scroll
    threshold     : 100       // DEFAULT: 0
});
```

####responsive
This is only needed by browsers that don't support CSS3. To accomplish responsive effect on older browsers, this plugin needs to do a re-calculation when `$(window).resize()` event is fired. The higher the number is the slower thumbnail gets re-calculated. 0 (zero) disables responsive feature in older browsers. **`modern` method does not support disabling responsive feature, use `method :"native"` would disable it.**
```javascript
/* responsive only works for native method / older browsers */
$('img').jqthumb({
    responsive : 10 // DEFAULT: 20
});

/* to disable responsive feature in modern method / browsers, switch method to native */
$('img').jqthumb({
    method     : 'native', // DEFAULT: auto
    responsive : 0         // DEFAULT: 20
});
```

####zoom
To zoom-in and out the thumbnail.
```javascript
$('img').jqthumb({
    zoom : 3 // DEFAULT: 1
});
```

####method
This plugin was built in two methods which one is for browsers that support CSS3 and another one is a native method that is fully done in mathematical calculation and it's for older browsers like IE6+ and browsers that don't support CSS3. Either one would have an identical result. In some cases, you might want to change the method to test or whatever. By default, the plugin detects your browsers compatability and assign method accordingly.
```javascript
$('img').jqthumb({
    method : 'native' // Availability: "auto", "modern", "native". DEFAULT: auto
});
```

####reinit
This lets the plugin know what to do when an image is intialized for the second time. By default, reinitiallization is enabled and the image will be killed and re-rendered. If set to `false`, nothing would happen.
```javascript
$('img').jqthumb({
    reinit : true // Availability: true / false. DEFAULT: true
});
```

####error
This callback returns the DOM and its image URL when error occurs.
```javascript
$('img').jqthumb({
    error : function(dom, imgUrl){
        console.log(dom, ' with its url "' + imgUrl + '" is invalid.');
    }
});
```

####before
This is a callback function which will be called right before calculation started. This function returns the original image source/object as its parameter. If you initialize the plugins with multiple-objects classname then this would be called for multiple times.
```javascript
$('img').jqthumb({
    before : function(originalImage){
        console.log(originalImage);
    }
});
```

####after
This is a callback function which will be called after everything is finished. This function returns the new generated thumbnail object as its parameter. If you initialize the plugin with multiple-objects classname then this would be called for multiple times.
```javascript
$('img').jqthumb({
    after : function(newThumb){
        $(newThumb).fadeIn();
    }
});
```

####done
This is a callback function which will be called when all objects have finished processing in a single plugin initialization. This returns an array type parameter that contains the object of all generated thumbnails.
```javascript
$('img').jqthumb({
    done : function(thumbnails){
        for(i in thumbnails)
            $(thumbnails[i]).fadeIn();
    }
});
```

##COMMANDS
```javascript
$('img').jqthumb('kill'); // destroy the plugin
$.jqthumb('killall');     // destroy all generated thumbnails on the page
```

##SEO IMPACT
You might be worried the SEO impact if you were to use this plugin. Maybe thought that changing `<img src="http://example.com/picture.jpg"/>` to `<img attr-src="http://example.com/picture.jpg"/>` would probably cause search engines not being able to crawl the images. Yes, it's right but not completely. You can always your `<noscript><img src="http://example.com/picture.jpg"/></noscript>` to output the image for search engines. Here are two examples:
```html
...
<img class="example" attr-src="http://example.com/picture.jpg" />
<noscript>
    <img src="http://example.com/picture.jpg" />
</noscript>
...
<script type="text/javascript">
    $(function(){
        $('.example[attr-src]').jqthumb({
            width  : 300,
            height : 200
        });
    });
</script>
```
Or a simplified version:
```html
...
<noscript attr-src="http://example.com/picture.jpg">
    <img src="http://example.com/picture.jpg" />
</noscript>
...
<script type="text/javascript">
    $(function(){
        $('noscript[attr-src]').jqthumb({
            width  : 300,
            height : 200
        });
    });
</script>
```

##TESTED BROWSERS
1. Google Chrome
2. Mozilla Firefox
3. Safari
4. Internet Explorer 6, 7, 8, 9 and 10

##EVEN MORE SAMPLE USAGE
```html
...
<img src="path/image.jpg" />
...
<script type="text/javascript">
    $(function(){
        $('img').jqthumb({
            width  : 300,
            height : 200
        });
    });
</script>
```

```html
...
<div data-jqthumb-src="path/image.jpg"></div>
...
<script type="text/javascript">
    $(function(){
        $('div').jqthumb({
            source : 'data-jqthumb-src'
        });
    });
</script>
```

```html
...
<div style="width: 100%; height:500px;">
    <img src="path/image.png" />
</div>
...
<script type="text/javascript">
    $(function(){
        $('div').jqthumb({
            width  : '100%',
            height : '100%'
        });
    });
</script>
```

```html
...
<img class="my-img" data-jqthumb-src="path/image1.png" data-jqthumb-width="200" data-jqthumb-height="200" />
<img class="my-img" data-jqthumb-src="path/image2.png" data-jqthumb-width="200" data-jqthumb-height="180" />
<img class="my-img" data-jqthumb-src="path/image3.png" data-jqthumb-width="200" data-jqthumb-height="160" />
<img class="my-img" data-jqthumb-src="path/image4.png" data-jqthumb-width="200" data-jqthumb-height="140" />
<img class="my-img" data-jqthumb-src="path/image5.png" data-jqthumb-width="200" data-jqthumb-height="120" />
...
<script type="text/javascript">
    $(function(){
        $('.my-img').each(function(){
            var $img = $(this);
            $img.jqthumb({
                source : $img.attr('data-jqthumb-src'),
                width  : $img.attr('data-jqthumb-width'),
                height : $img.attr('data-jqthumb-height')
            });
        });
    });
</script>
```

```html
...
<img class="my-img" src="path/image.jpg" />
...
<script type="text/javascript">
    $(function(){
        $('.my-img').jqthumb({
            width  : 300,
            height : 300,
            show   : false, // By default the image would be shown immediately after processing. To disable, set it to false
            after  : function(croppedImg){ // This callback returns an object
                $(croppedImg).fadeIn(); // This would fade in the cropped image
            }
        });
    });
</script>
```

```html
...
<img class="my-img" src="path/image1.jpg" />
<img class="my-img" src="path/image2.jpg" />
<img class="my-img" src="path/image3.jpg" />
...
<script type="text/javascript">
    $(function(){
        $('.my-img').jqthumb({
            width  : 300,
            height : 300,
            show   : false, // By default the image would be shown immediately after processing. To disable, set it to false
            done   : function(allCroppedImgs){ // This callback returns an array
                for(i in allCroppedImgs){

CHANGELOG.txt
Change Log:


# v1.1
===================================================
1. Uses jquery boilerplate written in https://github.com/jquery-boilerplate/jquery-boilerplate/blob/master/src/jquery.boilerplate.js as a plugin template
2. Dimension now supports percentage instead of fixed value. But it only works in CSS3-supported browsers.
3. Fixed bugs that occur when there are two plugin initializations and options in the 2nd one would replace the options in the first one. But I have no idea how I fixed it though, it just worked out of sudden after many hours of debugging.


# v1.2
===================================================
1. "complete" callback is now called "eachcomplete".
2. Added "allcomplete" callback to do only one single callback when everything is done.
3. Added global variables to store things to be used in every methods globally.
4. Added updateGlobal() method to update global variables and also keep all input and output elements.
5. Fixed bug - methods.getActualSize() has changed to this.getActualSize().


# v1.2.1
===================================================
1. Bug occurs when you have more than one initialzations and global.counter keep adding up from the previous initialzation. Bug fixed.


# v1.2.2
===================================================
1. Bug occurs when you have more than one initialzations and global.outputElems keep stacking up objects from the previous initialzation. Bug fixed.
2. Bug occurs in older browsers which causes images not showing up when "allcomplete" callback is used.
3. Removed unwanted parameter from contrustor.
4. Change of variable names.


# v1.3.0
===================================================
1. Added "img_src" attribute for specifying the image source. Because some people might want to do lazy-loading to optimize the performance of the site.


# v1.3.1
===================================================
1. Fixed browser-hang issue in IE8,7,6.
2. Added real demo image file instead of using a 64-bit image.


# v1.3.2
===================================================
1. Permanently fixed browser-hang issue in IE8,7,6.


# v1.3.3
===================================================
1. Width and height in percentage is now supported.
2. Split makethumbnail method into two different methods for better maintenance.
3. Added a local jQuery core file.


# v1.3.4
===================================================
1. This plugin now supports from jQuery v1.3 and above. Previously was v1.8 and above.


# v1.3.5
===================================================
1. Changed attribute name from "img_src" to "source".


# v1.4
===================================================
1. Added "position.top" and "position.left" attributes to allow users specify which portion of the photos to be shown.
2. Added "before" callback.
3. Renamed "eachcomplete" callback to "after".
4. Renamed "allcomplete" callback to "complete".
5. Simplified and fixed "percentOrPixel()" method.
6. Changed all "parseInt()" to "parseFloat()" to get more accurate values.
7. Replace 'px' from all atributes during plugin initialzation.


# v1.4.1
===================================================
1. Change callback name from "complete" to "done" as the name "complete" is a reserved by jQuery.


# v1.5
===================================================
1. Add <div/> image container wrap images, so the original element that wraps the image will not be modified in terms of styling.


# v1.6
===================================================
1. Add "kill" method to undo everything of the element.


# v1.6.1
===================================================
1. Not able to re-render image after killing it. Bug fixed.
2. Add plugin name as data to the element generated by jqthumb.
3. When killing, check the element inserted right before the targetted image has the plugin name as a data.


# v1.7.0
===================================================
1. Added Grunt to the project.
2. Fix potential errors and warnings reported by jshint.
3. Big change of the file structure. Demo files are in demo directory now.


# v1.7.3
===================================================
1. Fix "done" callback.
2. Change addBack() function to native for-loop method.
3. Start using Bower from now one.


# v1.8.0
===================================================
1. Support Zepto from now on
2. Fix "done" callback


# v1.9.0
===================================================
1. Added $.jqthumb() with the method "killall" to kill all rendered images


# v1.9.1 (Saturday, June 28th, 2014, 9:24:00 AM)
===================================================
1. Change Zepto version from 1.1.3 to 1.0.0
2. Change demo folder name to "src".
3. Better file structure in dist folder now.


# v1.9.2 (Saturday, June 28th, 2014, 9:35:37 AM)
===================================================
1. Better file structure in Gruntfile.js


# v1.9.3 (Thursday, July 10th, 2014, 4:32:44 PM)
===================================================
1. Added comments to Gruntfile.js
2. Updated jqthumb.jquery.json
3. Remove unwanted variables.


# v1.9.4 (Thursday, July 10th, 2014, 4:55:04 PM)
===================================================
1. Updated bower.json


# v1.9.5 (Friday, September 12th, 2014, 3:18:36 PM)
===================================================
1. Updated bower.json
2. Changed minimum zepta-data version to 1.0.0


# v1.9.6 (Thursday, September 18th, 2014, 9:27:33 PM)
===================================================
1. Fix background-position issue for css3 supported method


# v1.9.7 (Friday, September 19th, 2014, 2:01:49 PM)
===================================================
1. Re-design demo page and add playground


# v1.9.8 (Friday, September 19th, 2014, 3:40:23 PM)
===================================================
1. Fix console.log() in older browsers.
2. Fix screenshot taking area in Gruntfile.js.
3. Update readme structure in Gruntfile.js.
4. Update copy structure in Gruntfile.js.

jquery.thumb.js
/*
	Copyright (c) 2013-2014 Tho Pak Cheong
	Version: 1.6.1 (23-MAY-2014)
	Dual licensed under the MIT and GPL licenses.
	Requires: jQuery v1.3 or later

	Change Log:
	V1.1
	=========================
	1. Uses jquery boilerplate written in https://github.com/jquery-boilerplate/jquery-boilerplate/blob/master/src/jquery.boilerplate.js as a plugin template
	2. Dimension now supports percentage instead of fixed value. But it only works in CSS3-supported browsers.
	3. Fixed bugs that occur when there are two plugin initializations and options in the 2nd one would replace the options in the first one. But I have no idea how I fixed it though, it just worked out of sudden after many hours of debugging.

	V1.2
	=========================
	1. "complete" callback is now called "eachcomplete".
	2. Added "allcomplete" callback to do only one single callback when everything is done.
	3. Added global variables to store things to be used in every methods globally.
	4. Added updateGlobal() method to update global variables and also keep all input and output elements.
	5. Fixed bug - methods.getActualSize() has changed to this.getActualSize().

	V1.2.1
	=========================
	1. Bug occurs when you have more than one initialzations and global.counter keep adding up from the previous initialzation. Bug fixed.

	v1.2.2
	=========================
	1. Bug occurs when you have more than one initialzations and global.outputElems keep stacking up objects from the previous initialzation. Bug fixed.
	2. Bug occurs in older browsers which causes images not showing up when "allcomplete" callback is used.
	3. Removed unwanted parameter from contrustor.
	4. Change of variable names.

	v1.3.0
	=========================
	1. Added "img_src" attribute for specifying the image source. Because some people might want to do lazy-loading to optimize the performance of the site.

	v1.3.1
	=========================
	1. Fixed browser-hang issue in IE8,7,6.
	2. Added real demo image file instead of using a 64-bit image.

	v1.3.2
	=========================
	1. Permanently fixed browser-hang issue in IE8,7,6.

	v1.3.3
	=========================
	1. Width and height in percentage is now supported.
	2. Split makethumbnail method into two different methods for better maintenance.
	3. Added a local jQuery core file.

	v1.3.4
	=========================
	1. This plugin now supports from jQuery v1.3 and above. Previously was v1.8 and above.

	v1.3.5
	=========================
	1. Changed attribute name from "img_src" to "source".

	v1.4
	=========================
	1. Added "position.top" and "position.left" attributes to allow users specify which portion of the photos to be shown.
	2. Added "before" callback.
	3. Renamed "eachcomplete" callback to "after".
	4. Renamed "allcomplete" callback to "complete".
	5. Simplified and fixed "percentOrPixel()" method.
	6. Changed all "parseInt()" to "parseFloat()" to get more accurate values.
	7. Replace 'px' from all atributes during plugin initialzation.

	v1.4.1
	=========================
	1. Change callback name from "complete" to "done" as the name "complete" is a reserved by jQuery.

	v1.5
	=========================
	1. Add <div/> image container wrap images, so the original element that wraps the image will not be modified in terms of styling.

	v1.6
	=========================
	1. Add "kill" method to undo everything of the element.

	v1.6.1
	=========================
	1. Not able to re-render image after killing it. Bug fixed.
	2. Add plugin name as data to the element generated by jqthumb.
	3. When killing, check the element inserted right before the targetted image has the plugin name as a data.
*/

;(function ( $, window, document, undefined ) {

	var pluginName = "jqthumb",
		defaults = {
			classname: 'jqthumb',
			width: 100,
			height: 100,
			position: {
				top: '50%',
				left: '50%'
			},
			source: 'src',
			showoncomplete: true,
			before: function(){},
			after: function(){},
			done: function(){}
		},
		global = {
			elemCounter: 0,
			totalElems: 0,
			inputElems: {}, // store all the elements before executing
			outputElems: []
		};


	function Plugin ( element, options ) {// The actual plugin constructor
		this.element = element;
		this.settings = $.extend( {}, defaults, options );
			this.settings.width			= this.settings.width.toString().replace(/px/g, '');
			this.settings.height		= this.settings.height.toString().replace(/px/g, '');
			this.settings.position.top	= this.settings.position.top.toString().replace(/px/g, '');
			this.settings.position.left	= this.settings.position.left.toString().replace(/px/g, '');
		this._defaults = defaults;
		this._name = pluginName;

		if(typeof options == 'string'){
			if(options.toLowerCase() == 'kill'){
				this.kill(this.element);
			}
		}else{
			this.init();
		}
	}

	Plugin.prototype = {
		init: function () {
			if(this.support_css3_attr('backgroundSize') == false){ // old browsers need to do calculation to perform same output like "background-size: cover"
				this.nonCss3Supported_method(this.element, this.settings);
			}else{ // modern browsers that support CSS3 would be easier
				this.css3Supported_method(this.element, this.settings);
			}
		},

		kill: function(_this){
			if( $.data( _this, "plugin_" + pluginName )){

				if($(_this).prev().data('plugin') !== pluginName){
					console.error('We could not find the element created by jqthumb. It is probably due to one or more element has been added right before the image element after the plugin initialization, or it was removed.');
					return false;
				}

				$(_this).prev().remove();

				$(_this).removeAttr('style'); // first, remove all the styles first
				if(typeof $(_this).data('original-styles') !== 'undefined'){
					$(_this).attr('style', $(_this).data('original-styles')) // then re-store the original styles
				}

				if(typeof $(_this).data('original-styles') !== 'undefined'){
					$(_this).removeData('original-styles') // remove data that stores the original stylings before the image being rendered
				}

				if(typeof $(_this).data('plugin_' + pluginName) !== 'undefined'){
					$(_this).removeData('plugin_' + pluginName) // remove data that stored during plugin initialization
				}
			}
		},

		nonCss3Supported_method: function(_this, options){

			options.before.call(_this, _this);

			var that = this;

			$(_this).data('original-styles', $(_this).attr('style')); // keep original styles into data

			$(_this).hide();

			$(_this).one('load',function() {
				that.checkSrcAttrName(_this, options);

				var tempImg = $("<img/>").attr("src", $(_this).attr("src"));
				$(tempImg).load(function(){

					var newImg = {
							obj: tempImg,
							size: {
								width: this.width,
								height: this.height
							}
						},
						pw = that.percentOrPixel(options.width),
						ph = that.percentOrPixel(options.height),
						imgContainer = $('<div />');

					$(imgContainer)
						.insertBefore($(_this))
						.append($(newImg.obj))
						.css({
							'position'	: 'relative',
							'overflow'	: 'hidden',
							'width'		: (pw == '%') ? options.width : options.width + 'px',
							'height'	: (ph == '%') ? options.height : options.height + 'px'
						})
						.data('plugin', pluginName); // it would be easy to kill later

					if(newImg.size.width > newImg.size.height){ // horizontal

						$(newImg.obj).css({
							'width'			: 'auto',
							'max-height'	: 99999999,
							'min-height'	: 0,
							'max-width'		: 99999999,
							'min-width'		: 0,
							'height'		: $(newImg.obj).parent().height() + 'px'
						});

						var ratio = $(newImg.obj).height() / $(newImg.obj).width(); // get ratio

						if( $(newImg.obj).width() < $(newImg.obj).parent().width() ){
							$(newImg.obj).css({
								'width': $(newImg.obj).parent().width(),
								'height': parseFloat($(newImg.obj).parent().width() * ratio)
							});
						}

					}else{ // vertical

						$(newImg.obj).css({
							'width'			: $(newImg.obj).parent().width() + 'px',
							'max-height'	: 99999999,
							'min-height'	: 0,
							'max-width'		: 99999999,
							'min-width'		: 0,
							'height'		: 'auto'
						});

						var ratio = $(newImg.obj).width() / $(newImg.obj).height(); // get ratio

						if( $(newImg.obj).height() < $(newImg.obj).parent().height() ){
							$(newImg.obj).css({
								'width': parseFloat($(newImg.obj).parent().height() * ratio),
								'height': $(newImg.obj).parent().height()
							});
						}
					}

					posTop = (that.percentOrPixel(options.position.top) == '%') ? options.position.top : options.position.top + 'px';
					posLeft = (that.percentOrPixel(options.position.left) == '%') ? options.position.left : options.position.left + 'px';

					$(newImg.obj).css({
						'position'		: 'absolute',
						'top'			: posTop,
						'margin-top'	: function(){
							if(that.percentOrPixel(options.position.top) == '%'){
								return '-' + parseFloat(($(newImg.obj).height() / 100) * options.position.top.slice(0,-1)) + 'px'
							}
						},
						'left'			: posLeft,
						'margin-left'	: function(){
							if(that.percentOrPixel(options.position.left) == '%'){
								return '-' + parseFloat(($(newImg.obj).width() / 100) * options.position.left.slice(0,-1)) + 'px'
							}
						}
					});

					$(imgContainer)
						.hide()
						.addClass(options.classname);

					if(options.showoncomplete == true){
						$(imgContainer).show();
					}
					options.after.call(_this, $(imgContainer));

					that.updateGlobal(_this, $(imgContainer), options);

				}).each(function(){

					$(tempImg).load();

				});
			}).each(function(){

				$(_this).load();

			});
		},

		css3Supported_method: function (_this, options) {

			options.before.call(_this, _this);

			var that = this;

			$(_this).data('original-styles', $(_this).attr('style')); // keep original styles into data

			$(_this).hide();

			$(_this).one('load',function() {
				var pw = that.percentOrPixel(options.width),
					ph = that.percentOrPixel(options.height),
					featuredBgImgContainer,
					featuredBgImg;

				featuredBgImgContainer = $('<div/>')
											.css({
												'width'   : (pw == '%') ? options.width : options.width + 'px',
												'height'  : (ph == '%') ? options.height : options.height + 'px',
												'display' : 'none'
											})
											.addClass(options.classname)
											.data('plugin', pluginName); // it would be easy to kill later

				featuredBgImg = $('<div/>').css({
					'width'              : '100%',
					'height'             : '100%',
					'background-image'   : 'url("' + $(_this).attr(options.source) + '")',
					'-ms-filter'         : '"progid:DXImageTransform.Microsoft.AlphaImageLoader(src="'+$(_this).attr(options.source)+'",sizingMethod="scale")',
					'background-repeat'  : 'no-repeat',
					'background-position': (function(){
						var posTop = (that.percentOrPixel(options.position.top) == '%') ? options.position.top : options.position.top + 'px',
							posLeft = (that.percentOrPixel(options.position.left) == '%') ? options.position.left : options.position.left + 'px';
						return posTop + ' ' + posLeft;
					})(),
					'background-size'    : 'cover'
				})
				.appendTo($(featuredBgImgContainer));

				$(featuredBgImgContainer).insertBefore($(_this));

				if(options.showoncomplete == true){
					$(featuredBgImgContainer).show();
				}

				that.checkSrcAttrName(_this, options);

				options.after.call(_this, $(featuredBgImgContainer));

				that.updateGlobal(_this, $(featuredBgImgContainer), options);

			}).each(function(){

				$(_this).load();

			});
		},

		updateGlobal: function(_this, obj, options){
			global.outputElems.push($(obj));
			global.elemCounter = global.elemCounter + 1;
			if(global.elemCounter == global.totalElems){
				options.done.call(_this, global.outputElems);
			}
		},

		checkSrcAttrName: function(_this, options){
			if(
				options.source != 'src' &&
				(
					typeof $(_this).attr('src') == 'undefined' ||
					$(_this).attr('src') == ''
				)
			)
			{
				$(_this).attr('src', $(_this).attr(options.source));
			}
		},

		percentOrPixel: function(str){
			str = str.toString();
			if(str.match("px$") || str.match("PX$") || str.match("pX$") || str.match("Px$")) {
				return 'px';
			}else if(str.match("%$")) {
				return '%';
			}
		},

		support_css3_attr: (function() {
			/* code available at http://net.tutsplus.com/tutorials/html-css-techniques/quick-tip-detect-css-support-in-browsers-with-javascript/ */
			var div = document.createElement('div'),
				vendors = 'Khtml Ms O Moz Webkit'.split(' '),
				len = vendors.length;

			return function(prop) {
				if ( prop in div.style ) return true;

				prop = prop.replace(/^[a-z]/, function(val) {
					return val.toUpperCase();
				});

				for(i=0; i<vendors.length; i++){
					if ( vendors[i] + prop in div.style ) {
						return true;
					}
				}
				return false;
			};
		})()
	};

	$.fn[ pluginName ] = function ( options ) {

		if($.isFunction($.fn.addBack) == false){ // we need to use addBack functions which only exists from jQuery v1.3 and above.
			$.fn.extend({
				addBack: function( selector ) {
					return this.add( selector == null ?
						this.prevObject : this.prevObject.filter(selector)
					);
				}
			});
		}

		var elems = $(this).find('*').addBack();


		global.elemCounter = 0; // must always set to zero for every initialization
		global.outputElems = []; // must clear before doing anythong
		global.inputElems = $(elems);
		global.totalElems = $(elems).length; // set total of elements for later use.

		return this.each(function() {
			if(typeof options == 'string'){
				new Plugin( this, options );
			}else{
				if ( !$.data( this, "plugin_" + pluginName ) ) {
					$.data( this, "plugin_" + pluginName, new Plugin( this, options ) );
				}else{ // re-rendered without killing it
					new Plugin( this, 'kill' );
					$.data( this, "plugin_" + pluginName, new Plugin( this, options ) );
				}
			}
		});
	};

})( jQuery, window, document );